proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,175 @@
+which is not acceptable.
+
+We aim to automate the finding of highest optimistically confirmed slot and
+block data distribution, so that we can lower the possibility of human mistakes
+in the cluster restart process.
+
+## Alternatives Considered
+
+See [Handling Common Solana Outages]
+(https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc)
+for details.
+
+There are many proposals about automatically detecting that the cluster is
+in an outage so validators should enter a recovery process automatically.
+
+While getting human out of the loop greatly improves recovery speed,
+automaticlly restarting the whole cluster still seems risky. Because if
+the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+## New Terminology
+
+* SilentRepairMode - when validators restart in this new mode, they will
+talk with each other to find the highest optimistically confirmed slot and
+repair all necessary blocks. To improve speed and guarantee simplicity,
+Turbine, vote, and new block generation are paused in this mode.
+
+## Detailed Design
+
+The new protocol tries to make all 80% restarting validators get the same
+data blocks and the same set of last votes among them, then they can probably
+make the same decision and then proceed.
+
+The steps roughly look like this:
+
+1. Everyone freezes, no new blocks, no new votes, and no Turbine
+
+2. Make all blocks which can potentially have been optimistically confirmed
+before the freeze propagate to everyone
+
+3. Make restart participants' last votes before the freeze propagate to
+everyone
+
+4. Now see if enough people can optimistically agree on one block (same slot
+and hash) to restart from
+
+4.1 If yes, proceed and restart
+
+4.2 If no, freeze and print out what you think is wrong, wait for human
+
+A new command line arg --RepairAndRestart <optional_slots_to_send> is added.
+When the cluster is in need of a restart, we assume at least 80% will restart
+with this arg. Any validators restarted with this arg does not participate in
+the normal Turbine protocol, update its vote, or generate new blocks until all
+of the following steps are completed.
+
+### Gossip last vote before the restart and ancestors on that fork
+
+Send Gossip message LastVotedForkSlots to everyone, it contains the last voted
+slot on its tower and the ancestor slots on the last voted fork and is sent in
+a bitmap like the EpicSlots data structure. The number of ancestor slots sent
+is determined by <optional_slots_to_send_on_last_voted_fork>. By default this
+number is 108000, because that's 400ms * 10800 = 12 hours, we assume most
+restart decisions to be made in half a day. You can increase this number if you
+restart after the outage lasted more than 12 hours.
+
+### Aggregate, repair, and replay the slots in LastVotedForkSlots
```

Comment:
Eventually we'll want to describe the layout of the new `LastVotedForkSlots` and `Heaviest` gossip data values in this doc. Until they are fully formed it would be useful to have a placeholder at least

Reply:
Added.